velox/external/utf8proc/utf8procImpl.h

```diff
@@ -1007,6 +1007,102 @@ UTF8PROC_DLLEXPORT int utf8proc_codepoint_length(utf8proc_int32_t uc) {
     return 0;
 }
 
+/// This function is not part of the original utf8proc.
+/// Tries to get the length of UTF-8 encoded code point.  A
+/// positive return value means the UTF-8 sequence is valid, and
+/// the result is the length of the code point.  A negative return value means
+/// the UTF-8 sequence at the position is invalid, and the length of the invalid
+/// sequence is the absolute value of the result.
+///
+/// @param input Pointer to the first byte of the code point.
+/// @param size Number of available bytes.
+/// @return the length of the code point or negative the number of bytes in the
+/// invalid UTF-8 sequence.
+///
+/// Adapted from tryGetCodePointAt in
+/// https://github.com/airlift/slice/blob/master/src/main/java/io/airlift/slice/SliceUtf8.java
+UTF8PROC_DLLEXPORT int32_t
+utf8proc_try_char_length(const char* input, uint32_t size) {
+  auto charLength = utf8proc_char_length(input);
+  if (charLength < 0) {
+    return -1;
+  }
+
+  if (charLength == 1) {
+    // Normal ASCII: 0xxx_xxxx.
+    return 1;
+  }
+
+  auto firstByte = input[0];
+
+  // Process second byte.
+  if (size < 2) {
+    return -1;
+  }
+
+  auto secondByte = input[1];
+  if (!utf_cont(secondByte)) {
+    return -1;
+  }
+
+  if (charLength == 2) {
+    // 110x_xxxx 10xx_xxxx
+    int codePoint = ((firstByte & 0b00011111) << 6) | (secondByte & 0b00111111);
+    // Fail if overlong encoding.
+    return codePoint < 0x80 ? -2 : 2;
+  }
+
+  // Process third byte.
+  if (size < 3) {
+    return -2;
+  }
+
+  auto thirdByte = input[2];
+  if (!utf_cont(thirdByte)) {
+    return -2;
+  }
+
+  if (charLength == 3) {
+    // 1110_xxxx 10xx_xxxx 10xx_xxxx
+    int codePoint = ((firstByte & 0b00001111) << 12) |
+        ((secondByte & 0b00111111) << 6) | (thirdByte & 0b00111111);
+
+    // Surrogates are invalid.
+    static const int kMinSurrogate = 0xd800;
+    static const int kMaxSurrogate = 0xdfff;
+    if (kMinSurrogate <= codePoint && codePoint <= kMaxSurrogate) {
+      return -3;
+    }
+
+    // Fail if overlong encoding.
+    return codePoint < 0x800 ? -3 : 3;
+  }
+
+  // Process forth byte.
+  if (size < 4) {
+    return -3;
+  }
+
+  auto forthByte = input[3];
+  if (!utf_cont(forthByte)) {
+    return -3;
+  }
+
+  if (charLength == 4) {
+    // 1111_0xxx 10xx_xxxx 10xx_xxxx 10xx_xxxx
+    int codePoint = ((firstByte & 0b00000111) << 18) |
+        ((secondByte & 0b00111111) << 12) | ((thirdByte & 0b00111111) << 6) |
+        (forthByte & 0b00111111);
+    // Fail if overlong encoding or above upper bound of Unicode.
+    if (codePoint < 0x110000 && codePoint >= 0x10000) {
+      return 4;
+    }
+    return -4;
+  }
+
+  assert(false);
```

Comment:
in which cases do we hit this? If this is just a placeholder to make gcc happy, maybe we could re-organize the code to prevent this from being required?

Reply:
This only happens when `utf8proc_char_length` has a bug

Reply:
Just a placeholder to make gcc happy. Not sure how to reorganize the code to avoid this. Let me know if you have an idea.

Reply:
I think this is ok. The alternative could be that if the byte count is larger than 3, you assume it's 4 and return either 4 or -4 (without checking if it exceeds this boundary).